modules/programs/firefox/mkFirefoxModule.nix

```diff
@@ -0,0 +1,984 @@
+        }</H3>
+        ${indent indentLevel}<DL><p>
+        ${allItemsToHTML (indentLevel + 1) directory.bookmarks}
+        ${indent indentLevel}</DL><p>'';
+
+      itemToHTMLOrRecurse = indentLevel: item:
+        if item ? "url" then
+          bookmarkToHTML indentLevel item
+        else
+          directoryToHTML indentLevel item;
+
+      allItemsToHTML = indentLevel: bookmarks:
+        lib.concatStringsSep "\n"
+        (map (itemToHTMLOrRecurse indentLevel) bookmarks);
+
+      bookmarkEntries = allItemsToHTML 1 bookmarks;
+    in pkgs.writeText "${packageName}-bookmarks.html" ''
+      <!DOCTYPE NETSCAPE-Bookmark-file-1>
+      <!-- This is an automatically generated file.
+        It will be read and overwritten.
+        DO NOT EDIT! -->
+      <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">
+      <TITLE>Bookmarks</TITLE>
+      <H1>Bookmarks Menu</H1>
+      <DL><p>
+      ${bookmarkEntries}
+      </DL>
+    '';
+
+  mkNoDuplicateAssertion = entities: entityKind:
+    (let
+      # Return an attribute set with entity IDs as keys and a list of
+      # entity names with corresponding ID as value. An ID is present in
+      # the result only if more than one entity has it. The argument
+      # entities is a list of AttrSet of one id/name pair.
+      findDuplicateIds = entities:
+        filterAttrs (_entityId: entityNames: length entityNames != 1)
+        (zipAttrs entities);
+
+      duplicates = findDuplicateIds (mapAttrsToList
+        (entityName: entity: { "${toString entity.id}" = entityName; })
+        entities);
+
+      mkMsg = entityId: entityNames:
+        "  - ID ${entityId} is used by " + concatStringsSep ", " entityNames;
+    in {
+      assertion = duplicates == { };
+      message = ''
+        Must not have a ${name} ${entityKind} with an existing ID but
+      '' + concatStringsSep "\n" (mapAttrsToList mkMsg duplicates);
+    });
+
+  wrapPackage = package:
+    let
+      # The configuration expected by the Firefox wrapper.
+      fcfg = { enableGnomeExtensions = cfg.enableGnomeExtensions; };
+
+      # A bit of hackery to force a config into the wrapper.
+      browserName =
+        package.browserName or (builtins.parseDrvName package.name).name;
+
+      # The configuration expected by the Firefox wrapper builder.
+      bcfg = setAttrByPath [ browserName ] fcfg;
+
+    in if package == null then
+      null
+    else if isDarwin then
+      package
+    else if isWrapped then
+      package.override (old: {
+        cfg = old.cfg or { } // fcfg;
+        extraPolicies = (old.extraPolicies or { }) // cfg.policies;
+      })
+    else
+      (pkgs.wrapFirefox.override { config = bcfg; }) package { };
+
+in {
+  options = setAttrByPath modulePath {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
+      description = ''
+        Whether to enable ${cfg.name}.${
```

Comment:
I'm afraid that `${cfg.name}` cannot be used in descriptions. This is because the description is included in the manual and including `cfg.name` then causes the manual to have a dependency on the configuration. This, in turn, makes it more or less impossible to get the manual from a binary cache 🙁

The most convenient solution would probably be to grab the name from the first `mkFirefoxModule` argument.

Reply:
I tested the manual when I commited this and it worked just fine using the default name value. Wouldn't bet on this however, so I changed it nonetheless. The internal attribute is kept so that tests can access it.